### PR TITLE
VM Metadata API: Adds an option to disable VM metadata API call

### LINKED ISSUE #4187: Skip VmMetadataApiHandler when running outside a VM
**Describe the bug**

CosmosDB client tries to obtain VM info which only exists when running the service on Azure VMs

Either:
- Automatically detect it and don't try to send a request
- Offer a way to disable this handler when running the client in a different environment (e.g. locally)

```
Scope: HttpMethod:GET Uri:http://169.254.169.254/metadata/instance?api-version=2020-06-01 {OriginalFormat}:HTTP {HttpMethod} {Uri}
2023-11-15 15:07:07.241 (err ) GET 169.254.169.254//metadata/instance 

0: Exception: Connection refused (169.254.169.254:80) (System.Net.Http.HttpRequestException)
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.R9.Extensions.HttpClient.Logging.Internal.HttpLoggingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
0:0: Exception: Connection refused (System.Net.Sockets.SocketException)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|277_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)

```

**To Reproduce**
Steps to reproduce the behavior. If you can include code snippets or links to repositories containing a repro of the issue that can helps us in detecting the scenario it would speed up the resolution.

Run an app with CosmosDB client enabled and log configured outside of Azure

**Expected behavior**
Either:
- Automatically detect it and don't try to send a request
- Offer a way to disable this handler when running the client in a different environment (e.g. locally)

**Actual behavior**
Exception is thrown in the logs

**Environment summary**
SDK Version: 3.33.0
OS Version (e.g. Windows, Linux, MacOSX) macOS 14.1.1 (23B81)

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description
As part of this PR, adding capability to stop instance metadata API call
1. Introduce COSMOS_DISABLE_IMDS_ACCESS env variable, to stop calling Instance Metadata API call, if set as true.

and 2 minor variable name change (spelling correction)
## Type of change
- [] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #4187

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs [hunk 1]
```diff
@@ -34,28 +34,43 @@ internal static class VmMetadataApiHandler
...
-                return;
+                isVMMetadataAccessDisabled = ConfigurationManager.GetEnvironmentVariable<bool>("COSMOS_DISABLE_VM_METADATA_ACCESS", false);
+            }
+            catch (Exception ex)
+            {
+                DefaultTrace.TraceWarning($"Error while reading environment variable COSMOS_DISABLE_VM_METADATA_ACCESS {0}", ex.Message);
             }
 
-            lock (VmMetadataApiHandler.lockObject)
+            if (System.Diagnostics.Debugger.IsAttached || isVMMetadataAccessDisabled)
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:49]: I think whether the debugger is attached or not should only change the default value - that way env variable can still be used to force the VM instance metadata under debugger in case you actually want to debug it ?

File: Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs [hunk 2]
```diff
@@ -34,28 +34,43 @@ internal static class VmMetadataApiHandler
...
         private static AzureVMMetadata azMetadata = null;
-       
         internal static void TryInitialize(CosmosHttpClient httpClient)
         {
-            if (VmMetadataApiHandler.isInitialized)
+            bool isVMMetadataAccessDisabled = false; // Default value
+            try
             {
-                return;
+                isVMMetadataAccessDisabled = ConfigurationManager.GetEnvironmentVariable<bool>("COSMOS_DISABLE_VM_METADATA_ACCESS", false);
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:42]: Looking at other places where we call GetEnvironmentVariable i don't see any special-casing of exceptions - is this try/catch really needed? What exception would we possibly expect?

Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:42]: My thought behind that was, if somebody is messing with this env variable (by passing invalid value
), then just fall back to default behavior.
But drawback is, it might confuse some people. so on second thought, removing this try-catch and let it throw the env related exception. Client initialization won't be possible with invalid of this env variable.

File: Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs [hunk 3]
```diff
@@ -34,28 +34,45 @@ internal static class VmMetadataApiHandler
...
-            if (VmMetadataApiHandler.isInitialized)
+            bool isVMMetadataAccessDisabled = 
+                ConfigurationManager.GetEnvironmentVariable<bool>("COSMOS_DISABLE_VM_METADATA_ACCESS", System.Diagnostics.Debugger.IsAttached);
+            if (isVMMetadataAccessDisabled)
             {
                 return;
             }
-
-            lock (VmMetadataApiHandler.lockObject)
+            else
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:55]: you don't need the else

**Review by FabianMeiswinkel — DISMISSED** [review]
LGTM except one nit.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs [hunk 4]
```diff
@@ -53,13 +56,28 @@ static Task<HttpResponseMessage> sendFunc(HttpRequestMessage request, Cancellati
...
+
+                Environment.SetEnvironmentVariable("COSMOS_DISABLE_VM_METADATA_ACCESS", "false");
+            }
+            else
+            {
+                Assert.AreEqual($"{VmMetadataApiHandler.VmIdPrefix}{"d0cb93eb-214b-4c2b-bd3d-cc93e90d9efd"}", VmMetadataApiHandler.GetMachineId());
+                Assert.AreEqual(VmMetadataApiHandler.GetMachineRegion(), "eastus");
+            }
+
+            Environment.SetEnvironmentVariable("COSMOS_DISABLE_VM_METADATA_ACCESS", null);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs:80]: Do this on a try/finally, otherwise if any of the Assert fails, you affected the machine.

```
Environment.SetEnvironmentVariable("COSMOS_DISABLE_VM_METADATA_ACCESS", isVmMetadataAccessDisabled);
try
{
  // test code
}
finally
{
  Environment.SetEnvironmentVariable("COSMOS_DISABLE_VM_METADATA_ACCESS", null);
}
```

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs [hunk 5]
```diff
@@ -53,13 +56,28 @@ static Task<HttpResponseMessage> sendFunc(HttpRequestMessage request, Cancellati
...
+            CosmosHttpClient cosmosHttpClient = MockCosmosUtil.CreateCosmosHttpClient(() => new HttpClient(messageHandler));
 
-            VmMetadataApiHandler.TryInitialize(cosmoshttpClient);
+            VmMetadataApiHandler.TryInitialize(cosmosHttpClient);
 
             await Task.Delay(2000);
-            Assert.AreEqual($"{VmMetadataApiHandler.VmIdPrefix}{"d0cb93eb-214b-4c2b-bd3d-cc93e90d9efd"}", VmMetadataApiHandler.GetMachineId());
-            Assert.AreEqual(VmMetadataApiHandler.GetMachineRegion(), "eastus");
+
+            if (isVmMetadataAccessDisabled != null &&
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs:65]: Based on the `DataRow` isVmMetadataAccessDisabled will never be null?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs [hunk 6]
```diff
@@ -53,13 +56,28 @@ static Task<HttpResponseMessage> sendFunc(HttpRequestMessage request, Cancellati
...
-            Assert.AreEqual(VmMetadataApiHandler.GetMachineRegion(), "eastus");
+
+            if (isVmMetadataAccessDisabled != null && 
+                Boolean.TryParse(isVmMetadataAccessDisabled, out bool isVmMetadataAccessDisabledBool) && 
+                isVmMetadataAccessDisabledBool)
+            {
+                Assert.AreNotEqual($"{VmMetadataApiHandler.VmIdPrefix}{"d0cb93eb-214b-4c2b-bd3d-cc93e90d9efd"}", VmMetadataApiHandler.GetMachineId());
+                Assert.IsNull(VmMetadataApiHandler.GetMachineRegion(), VmMetadataApiHandler.GetMachineRegion());
+
+                Environment.SetEnvironmentVariable("COSMOS_DISABLE_VM_METADATA_ACCESS", "false");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs:72]: WHy is this needed? You are clearing up the variable at the end

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs (hunk 5 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/VmMetadataApiHandlerTest.cs]: Yes..if it is null then, default value will be returned i.e. true/false based on debugger attached.

**Review by ealsur — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs [hunk 7]
```diff
@@ -34,9 +34,25 @@ internal static class VmMetadataApiHandler
...
+        /// If environment variable is set to true, then VM metadata call will not be made.
+        /// If environment variable is set to false, then VM metadata call will be made.
+        /// If environment variable is not set, then set the value according to the debugger which means 
+        ///     In Debug mode, it is will be true and VM metadata call will be disabled.
+        ///     In Release mode, it is will be false and VM metadata call will be enabled.
+        /// </summary>.
+        /// <param name="httpClient"></param>
         internal static void TryInitialize(CosmosHttpClient httpClient)
         {
+            bool isVMMetadataAccessDisabled =
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:49]: `System.Diagnostics.Debugger.IsAttached` is this intentional?

Non-deterministic for customers?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:49]: Ex: Azure app services do allow attaching debugger (i.e. in the cloud environment)

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:49]: Two side-affects
- Remote debugging (ex: App services) might run into it (also it needs to be at bootstrap time)
- With and with-out debugging in local debug in-consistency 

Later is more important than first. 
Inconsistency seems like a conscious choice here

Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:47]: Removed debugger logic, now env variable is the only way to control this call.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs [hunk 8]
```diff
@@ -34,9 +34,23 @@ internal static class VmMetadataApiHandler
...
+        /// Check for environment variable COSMOS_DISABLE_VM_METADATA_ACCESS to decide if VM metadata call should be made or not.
+        /// If environment variable is set to true, then VM metadata call will not be made.
+        /// If environment variable is set to false, then VM metadata call will be made.
+        /// If environment variable is not set, then VM metadata call will be made.
+        /// </summary>.
+        /// <param name="httpClient"></param>
         internal static void TryInitialize(CosmosHttpClient httpClient)
         {
+            bool isVMMetadataAccessDisabled = 
+                ConfigurationManager.GetEnvironmentVariable<bool>("COSMOS_DISABLE_VM_METADATA_ACCESS", false);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:48]: Naming suggestion: COSMOS_DISABLE_IMDS_ACCESS

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Telemetry/VmMetadataApiHandler.cs:48]: Also can we please cover this in our known issue section? (for visibility)

**Review by kirankumarkolli — DISMISSED** [review]

**Comment by kirankumarkolli** [comment]
Please refresh the description with new ENV naming.

**Review by kirankumarkolli — APPROVED** [review]
